numpy/core/src/multiarray/alloc.c

```
@@ -433,10 +441,14 @@ PyDataMem_UserNEW_ZEROED(size_t nmemb, size_t size, const PyDataMemAllocator *al
 }
 
 NPY_NO_EXPORT void
-PyDataMem_UserFREE(void *ptr, size_t size, const PyDataMemAllocator *allocator)
+PyDataMem_UserFREE(void *ptr, size_t size, PyObject *mem_handler)
 {
+    PyDataMem_Handler *handler = (PyDataMem_Handler *) PyCapsule_GetPointer(mem_handler, "mem_handler");
+    if (handler == NULL) {
+        return;
```

Here `PyCapsule_GetPointer` may fail (level: paranoid). `PyDataMem_UserFREE` is currently `void`.

Should we propagate the error (e.g. `return -1`/`return 0`) changing the function's signature, returning `int`?

If yes, I guess that most of the functions that use it won't care about failed frees?
If no, `PyCapsule_GetPointer` always sets an exception on failure, should we `PyErr_Clear` it before returning?

---

Sorry for being so insecure about `PyErr`. I guess I have to start think of it more like `errno`, and let it be?

---

@eric-wieser what do you think? I guess this is the price of adding another layer of abstraction.

---

Free should probably use `PyErr_WriteUnraisable` if it fails, like `__del__` does.

(Note: I've only looked at the context not the full diff)